File: src/stores/encryption.ts
import { create } from "zustand";
import { api } from "../api";
import type { EncryptionStatus } from "../types";

interface EncryptionState {
  status: EncryptionStatus | null;
  loading: boolean;
  error: string | null;
  refresh: () => Promise<void>;
  clear: () => void;
  enable: (passphrase: string) => Promise<void>;
  unlock: (passphrase: string) => Promise<void>;
  lock: () => Promise<void>;
  changePassphrase: (oldPass: string, newPass: string) => Promise<void>;
  disable: (passphrase: string) => Promise<void>;
}

async function call<T>(
  set: (p: Partial<EncryptionState>) => void,
  fn: () => Promise<T>,
): Promise<T> {
  set({ loading: true, error: null });
  try {
    const res = await fn();
    set({ loading: false });
    return res;
  } catch (e) {
    set({ loading: false, error: String(e) });
    throw e;
  }
}

export const useEncryptionStore = create<EncryptionState>((set) => ({
  status: null,
  loading: false,
  error: null,

  clear: () => set({ status: null, error: null, loading: false }),

  refresh: async () => {
    const s = await call(set, () => api.encryptionStatus());
    set({ status: s });
  },

  enable: async (passphrase) => {
    const s = await call(set, () => api.enableEncryption(passphrase));
    set({ status: s });
  },

  unlock: async (passphrase) => {
    const s = await call(set, () => api.unlockEncryption(passphrase));
    set({ status: s });
  },

  lock: async () => {
    const s = await call(set, () => api.lockEncryption());
    set({ status: s });
  },

  changePassphrase: async (oldPass, newPass) => {
    const s = await call(set, () => api.changeEncryptionPassphrase(oldPass, newPass));
    set({ status: s });
  },

  disable: async (passphrase) => {
    const s = await call(set, () => api.disableEncryption(passphrase));
    set({ status: s });
  },
}));

// Convenience selector.
export const selectLocked = (s: EncryptionState): boolean =>
  !!s.status && s.status.enabled && !s.status.unlocked;

File: src-tauri/src/commands/encryption.rs
//! End-to-end encryption commands.

use serde::Serialize;
use tauri::State;

use crate::encryption::{self, EncryptionMeta};
use crate::error::{AppError, AppResult};
use crate::state::AppState;

#[derive(Debug, Clone, Serialize)]
pub struct EncryptionStatus {
    pub enabled: bool,
    pub unlocked: bool,
    pub meta: Option<EncryptionMeta>,
}

#[tauri::command]
pub fn encryption_status(state: State<'_, AppState>) -> AppResult<EncryptionStatus> {
    let g = state.current()?;
    let meta = encryption::load_meta(&g.encryption_meta_path)?;
    Ok(EncryptionStatus {
        enabled: meta.is_some(),
        unlocked: g.keyring.is_unlocked(),
        meta,
    })
}

#[tauri::command]
pub async fn enable_encryption(
    passphrase: String,
    state: State<'_, AppState>,
) -> AppResult<EncryptionStatus> {
    let g = state.current()?;
    if encryption::load_meta(&g.encryption_meta_path)?.is_some() {
        return Err(AppError::Other("encryption already enabled".into()));
    }
    let (key, meta) = encryption::enable(&g.encryption_meta_path, &passphrase)?;
    g.keyring.set(key);
    g.keyring.set_active(true);
    // Re-encrypt every existing block through the wrapped backend.
    let blocks = g.backend.all_blocks().await?;
    for b in blocks {
        g.backend.update_block(&b.id, &b.content).await?;
    }
    Ok(EncryptionStatus {
        enabled: true,
        unlocked: true,
        meta: Some(meta),
    })
}

#[tauri::command]
pub fn unlock_encryption(
    passphrase: String,
    state: State<'_, AppState>,
) -> AppResult<EncryptionStatus> {
    let g = state.current()?;
    let meta = encryption::load_meta(&g.encryption_meta_path)?
        .ok_or_else(|| AppError::Other("encryption not enabled".into()))?;
    let key = encryption::unlock(&meta, &passphrase)?;
    g.keyring.set(key);
    g.keyring.set_active(true);
    Ok(EncryptionStatus {
        enabled: true,
        unlocked: true,
        meta: Some(meta),
    })
}

#[tauri::command]
pub fn lock_encryption(state: State<'_, AppState>) -> AppResult<EncryptionStatus> {
    let g = state.current()?;
    g.keyring.clear();
    let meta = encryption::load_meta(&g.encryption_meta_path)?;
    Ok(EncryptionStatus {
        enabled: meta.is_some(),
        unlocked: false,
        meta,
    })
}

#[tauri::command]
pub async fn change_encryption_passphrase(
    old_passphrase: String,
    new_passphrase: String,
    state: State<'_, AppState>,
) -> AppResult<EncryptionStatus> {
    let g = state.current()?;
    if encryption::load_meta(&g.encryption_meta_path)?.is_none() {
        return Err(AppError::Other("encryption not enabled".into()));
    }
    // Ensure current vault is accessible before rotating.
    let meta = encryption::load_meta(&g.encryption_meta_path)?.unwrap();
    let _old_key = encryption::unlock(&meta, &old_passphrase)?;
    // Generate new meta (new salt + key). We must first decrypt all blocks
    // with the OLD key, then rotate, then re-encrypt with the new key.
    // The decorator is currently bound to `g.keyring`, which is mutable.
    // Decrypt-pass: temporarily ensure we're using the old key.
    g.keyring.set(_old_key);
    g.keyring.set_active(true);
    let plaintexts: Vec<(String, String)> = g
        .backend
        .all_blocks()
        .await?
        .into_iter()
        .map(|b| (b.id, b.content))
        .collect();
    // Rotate to new key.
    let (new_key, new_meta) =
        encryption::change_passphrase(&g.encryption_meta_path, &old_passphrase, &new_passphrase)?;
    g.keyring.set(new_key);
    g.keyring.set_active(true);
    for (id, plain) in plaintexts {
        g.backend.update_block(&id, &plain).await?;
    }
    Ok(EncryptionStatus {
        enabled: true,
        unlocked: true,
        meta: Some(new_meta),
    })
}

#[tauri::command]
pub async fn disable_encryption(
    passphrase: String,
    state: State<'_, AppState>,
) -> AppResult<EncryptionStatus> {
    let g = state.current()?;
    let meta = encryption::load_meta(&g.encryption_meta_path)?
        .ok_or_else(|| AppError::Other("encryption not enabled".into()))?;
    let key = encryption::unlock(&meta, &passphrase)?;
    g.keyring.set(key);
    g.keyring.set_active(true);
    // Decrypt every block and write plaintext back.
    let plaintexts: Vec<(String, String)> = g
        .backend
        .all_blocks()
        .await?
        .into_iter()
        .map(|b| (b.id, b.content))
        .collect();
    // Turn encryption off; now writes go through as plaintext.
    g.keyring.set_active(false);
    g.keyring.clear();
    for (id, plain) in plaintexts {
        g.backend.update_block(&id, &plain).await?;
    }
    encryption::disable(&g.encryption_meta_path)?;
    Ok(EncryptionStatus {
        enabled: false,
        unlocked: false,
        meta: None,
    })
}
